fw/tls.c

```diff
@@ -478,18 +484,58 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int mss_now,
 
 	spin_unlock(&tls->lock);
 
+free_pages:
 	for (p = pages; p < pages_end; ++p)
 		put_page(*p);
 
-out:
 	if (unlikely(sgt.nents > AUTO_SEGS_N))
 		kfree(sgt.sgl);
-	if (!r)
-		return r;
-err_epilogue:
-	T_WARN("%s: cannot encrypt data (%d), only partial data was sent\n",
-	       __func__, r);
+
+	if (likely(!r))
+		return 0;
+
+	if (unlikely(r == -ENOMEM)) {
+		struct sk_buff *tmp;
+		/* Restore TLS type, later we call this function again. */
+		for (tmp = skb; tmp != next; ) {
+			tmp = skb_queue_next(&sk->sk_write_queue, tmp);
+			skb_set_tfw_tls_type(tmp, type);
+		}
+	}
+
+restore_write_queue:
+	tcp_sk(sk)->write_seq -= head_sz + TTLS_TAG_LEN;
+	if (skb == skb_tail) {
+		BUG_ON(skb_was_added);
+		ss_skb_chop_head_tail_nofail(skb->next, skb, head_sz,
+					     TTLS_TAG_LEN);
+		TCP_SKB_CB(skb_tail)->end_seq -= TTLS_TAG_LEN;
+		tfw_tcp_propagate_dseq(sk, skb);
+	} else {
+		if (likely(!skb_was_added)) {
+			ss_skb_chop_head_tail_nofail(NULL, skb, head_sz, 0);
+			ss_skb_chop_head_tail_nofail(skb_tail->next, skb_tail,
+						     0, TTLS_TAG_LEN);
+			TCP_SKB_CB(skb_tail)->end_seq -= TTLS_TAG_LEN;
+		} else {
+			ss_skb_chop_head_tail_nofail(NULL, skb, head_sz, 0);
+			/*
+			 * New skb with TLS tag was added to socket write queue,
+			 * remove it.
+			 */
+			__skb_unlink(skb_tail, &sk->sk_write_queue);
+			skb_tail = skb_tail_save;
+		}
+		tfw_tcp_propagate_dseq(sk, skb);
+		tfw_tcp_propagate_dseq(sk, skb_tail);
```

Comment:
Why don't we reduce the skb's `end_seq` by the TLS header size to address this addition?
```C
/* TLS header is always allocated from the skb headroom. */
tcb->end_seq += head_sz;
```